components/datetime/src/format/time_zone.rs

```diff
@@ -51,8 +52,83 @@ where
     T: TimeZoneInput,
     W: fmt::Write + ?Sized,
 {
-    let pattern = &time_zone_format
-        .patterns
+    match (&time_zone_format.patterns, &time_zone_format.config) {
+        (Some(patterns), None) => write_pattern(time_zone_format, time_zone, patterns, w),
+        (None, Some(config)) => write_config(time_zone_format, time_zone, config, w),
+        (Some(_), Some(_)) => {
+            unreachable!("TimeZoneFormat should have a configuration or a pattern, but found both")
+        }
+        (None, None) => unreachable!(
+            "TimeZoneFormat should have either a configuration or a pattern, but found neither"
+        ),
```

Comment:
@sffc 

By your same logic about the `expect` potentially adding unwanted panic machinery, should we remove the `unreachable!()` calls here and return an error instead?

Reply:
I'm not quite as concerned about a plain `unreachable!()` because it's more self-contained; `.expect()` pulls in the formatting machinery since it prints a formatted message\*.  If we build with `panic_abort`, then any `unreachable!()` should basically compile away.

That being said, if you want to avoid the unreachable code path, you can use an internal enum such as

```rust
enum TimeZoneFormatKind {
  Config(TimeZoneFormatConfig),
  Pattern(DataPayload<'data, PatternPluralsFromPatternsV1Marker>),
}
```

This would likely be a good cleanup to your code regardless.

Reply:
I'm re-opening this thread because I think switching to use an enum may be a beneficial change for @samchen61661 to consider.

Reply:
Yeah, it is a good suggestion. I am not familiar with the open source policy. What do you think to improve it in a separate PR?

The main concern here is that this PR is big and the fix is relatively small.

Reply:
Got it. 

When you said "This would likely be a good cleanup to your code regardless." I thought you meant as a follow-up. 

Happy to add it in here to as I agree that it is a good change. 

@samchen61661 you can go ahead and add it to this PR.

Reply:
I'm fine landing this as-is and doing this refactor in a small follow-up PR.  But since you still need to resolve the merge conflict in this PR, it seems to make sense to fix this while you're at it.

Reply:
I see. sg and I will refactor it in this PR.

Reply:
Shane, when you talked about `you can use an INTERNAL exhaustive enum`, are you saying that I just need to use `TimeZoneFormatKind` in the `write_zone` method?

I think we can make `TimeZoneFormatKind` as an external attribute in `TimeZoneFormat` to replace `config` and `patterns`.

Reply:
@samchen61661 

What we would need to do is:

```rust
pub(super) TimeZoneFormatKind<'data> {
    ...
}
```

The enum should only be internal (i.e. `pub(super)`) because users of the public API shouldn't have to worry about this type. 

And then instead of having data members for `config` and `patterns`, you would just have one data member `kind` that is a `TimeZoneFormatKind`.

Note that the enum `TimeZoneFormatKind` will need to have a lifetime specifier since it holds the `DataPayload` for the pattern variant.